Add protection for disabled FEDs in the SiStrip packer

pieterdavid opened the pull request:
This PR fixes a problem found during the HI rehearsal (the packer tries to copy the FED header without checking that it sent any data, which leads to a segmentation fault at the repacking stage when there are disabled FEDs).
Most of the diff is due to whitespace changes (indentation) and moving the `WarningSummary` helper class (which prints only the first warning and keeps a count afterwards, unless debugging is enabled) from inside `SiStripRawToDigiUnpacker` to `EventFilter/SiStripRawToDigi/plugins`.

CC: @icali @erikbutz @mmusich

pieterdavid committed:
Add protection for disabled FEDS when repacking

and moved WarningSummary (print only the first warning by default
and a summary at endStream) out of SiStripRawToDigiUnpacker for
reuse in SiStripDigiToRaw

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24577/6469

cmsbuild commented:
A new Pull Request was created by  @pieterdavid (Pieter David) for master.

It involves the following packages:

EventFilter/SiStripRawToDigi


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald, @VinInn this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 19e622e49166d7b3749e1befcdb45fc7cf1dac43
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24577/30481/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/04963397facc38ce9efc0870e8c89f767673d402
https://github.com/cms-sw/cmssw/commit/55c368b77877ba6e8082b745b282782042cc4786
https://github.com/cms-sw/cmssw/commit/2f53b859736ea70295be75ed487d25fc6c343130
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24577/30481/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24577/30481/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24577/30481/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3146746
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3146546
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

venturia commented:
For my curiosity: why we observe this problem only now? It is not the first time we run with excluded FEDs

pieterdavid commented:
@venturia good question... this only shows up when repacking (and copying the FED headers from the input raw event), I think that was only used for heavy ion runs before (2015 in any case, earlier I'm not sure). Could it be that no FEDs were excluded during those?

venturia commented:
On 19 Sep 2018, at 11:01, Pieter David <notifications@github.com<mailto:notifications@github.com>> wrote:


@venturia<https://github.com/venturia> good question... this only shows up when repacking (and copying the FED headers from the input raw event), I think that was only used for heavy ion runs before (2015 in any case, earlier I'm not sure). Could it be that no FEDs were excluded during those?

maybe

       Thanks

           Andrea

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/24577#issuecomment-422718287>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEx-pyulIsWqHpNP7yIk5w-9yPgE3s11ks5ucggHgaJpZM4WuGpj>.

venturia commented:
Maybe at that time the FEDs were excluded and removed from the (offline) cabling map?

slava77 commented:
@cmsbuild please test workflow 140.55

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 19e622e49166d7b3749e1befcdb45fc7cf1dac43
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24577/30536/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24577/30536/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-24577/140.55_RunHI2015VR+RunHI2015VR+HYBRIDRepackHI2015VR+HYBRIDZSHI2015+RECOHID15+HARVESTDHI15

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3146746
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3146548
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

slava77 commented on EventFilter/SiStripRawToDigi/plugins/WarningSummary.cc:
```diff
@@ -0,0 +1,31 @@
+      [&message] ( const std::pair<std::string,std::size_t>& item ) { return item.first == message; } );
+  if ( std::end(m_warnings) == wIt ) {
+    m_warnings.emplace_back(message, 1);
+    edm::LogWarning(m_category) << message << ": " << details << (m_debug?"":"\nNote: further warnings of this type will be suppressed (this can be changed by enabling debugging printout)");
```
please add a newline for readbility

slava77 commented on EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc:
```diff
@@ -25,7 +25,8 @@ namespace sistrip {
     useFedKey_(useFedKey),
-    bufferGenerator_()
+    bufferGenerator_(),
+    warnings_("DigiToRaw", "[sistrip::DigiToRaw::createFedBuffers_]", edm::isDebugEnabled())
```
shouldn't this use a common constant from DataFormats/SiStripCommon/interface/ConstantsForLogger.h
similar to ```sistrip::RawToDigiUnpacker``` ?
... OTOH, it doesn't look like ```mlRawToDigi_``` is used consistently elsewhere with frequent direct ```"SiStripRawToDigi"```.
So, feel free to ignore.

pieterdavid replied:
I solved it like this to keep the more or less the same warning messages. There is no constant for the packer, and they are used rather inconsistently in several places... but I would leave cleaning that up for another time.

slava77 commented on EventFilter/SiStripRawToDigi/plugins/WarningSummary.h:
```diff
@@ -0,0 +1,24 @@
+#include <string>
+#include <vector>
+
+class WarningSummary {
```
the name is a bit too generic to be in the global namespace.
Please move to ```namespace sistrip```

slava77 commented on EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc:
```diff
@@ -145,176 +146,178 @@ namespace sistrip {
-	      << " - statusregister " << bufferGenerator_.trackerSpecialHeader().fedStatusRegister() << "\n"
-	      << " SpecialHeader: " << debugStream.str();
-	  }
+          if ( rawfedData.size() == 0 ) {
```
the code appears to be indented inconsistently. Please correct.

https://github.com/cms-sw/cmssw/blob/19e622e49166d7b3749e1befcdb45fc7cf1dac43/EventFilter/SiStripRawToDigi/plugins/SiStripDigiToRaw.cc#L147-L150

slava77 commented on EventFilter/SiStripRawToDigi/plugins/WarningSummary.h:
```diff
@@ -0,0 +1,24 @@
+  bool m_debug;
+  std::string m_category;
+  std::string m_name;
+  std::vector<std::pair<std::string,std::size_t>> m_warnings;
```
I wonder why this is not a map? access will be more efficient.
Is the insertion order important?

makortel replied:
I didn't take a look if it is the case here, but usually a sorted vector of (key, value) pairs has been faster for lookup than a map.

pieterdavid committed:
EventFilter/SiStripRawToDigi: WarningSummary updates

- moved to the sistrip namespace
- changed storage from vector<pair<string,size_t>> to map<string,size_t>
- long line split for readability

pieterdavid committed:
SiStripDigiToRaw: convert tabs to spaces

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24577/6512

cmsbuild commented:
Pull request #24577 was updated. @perrotta, @cmsbuild, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test workflow 140.55

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d2ef07fae6fe236db4c71afe6b349fd1abd05450
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24577/30548/summary.html

cmsbuild commented:
Comparison job queued.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/24577	d2ef07f
- code changes are in line with the PR description and the follow up review. 
- jenkins tests pass (no changes are expected in the tests working in the baseline).

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
